### ISSUE TITLE
Versions File

### ISSUE THREAD

**Issue by tmrobert8** [openhab/openhab-distro#1064]
Largely from a discussion with @kaikreuzer - the issue with the installation/update scripts is that the URLs for the various distributions (snapshots, milestones, stables) are hardcoded.  The creates thre issues;
1. If something deviates from established norm (such as the 2.5.x downloads), the scripts become unusable without manual changes on the user's end.
2. The scripts have no idea what versions are available and depend on the user to enter the version they want (whether it's available or not) and in the correct format (2.5.0-snapshot, etc).
3. If we ever change providers or URLs, the update scripts break until someone either manually makes a change or redownloads the newest script (with the updated URLs).

I'd like to propose a file in the distro repo that will describe all the various openHAB versions available, some metadata about the version (useful for the scripts) and the download URL for each.

That way the installation/update scripts (including openhabian) can, at runtime:
1.  Know the download URL for each version
2.  Be able to display a list of versions for the user to choose
3.  Have extra meta data for the version that the script can use (such as a sorting order, version category [stable, milestone, etc] - rather than relying on parsing the version number - allows the version numbers to vary in format if we want).

Although I really don't care what format the file should be (comma delimited, xml, json) - I'm going to suggest json...

The following could be an example file:
```
[
	{
		"version": "2.5.0-SNAPSHOT",
		"downloadUrl": "https://ci.openhab.org/job/openHAB-Distribution/lastSuccessfulBuild/artifact/distributions/openhab/target/openhab-2.5.0.zip",
		"addonUrl": "https://ci.openhab.org/job/openHAB-Distribution/lastSuccessfulBuild/artifact/distributions/openhab-addons/target/openhab-addons-2.5.0.kar",
		"meta": {
			"title" : "2.5.0 Snapshot"
			"category": "snapshot",
			"displayOrder": 51
		}
	},
	{
		"version": "2.5.0",
		"downloadUrl": "https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab%2F2.5.0%2Fopenhab-2.5.0.zip",
		"addonUrl": "https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab-addons%2F2.5.0%2Fopenhab-addons-2.5.0.kar",
		"meta": {
			"default" : true,
			"title" : "2.5.0 Stable"
			"category": "stable",
			"displayOrder": 50
		}
	}
	{
		"version": "2.4.0",
		"downloadUrl": "https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab%2F2.5.0%2Fopenhab-2.4.0.zip",
		"addonUrl": "https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab-addons%2F2.5.0%2Fopenhab-addons-2.4.0.kar",
		"meta": {
			"title" : "2.4.0 Stable"
			"category": "stable",
			"displayOrder": 49
		}
	}
	// every old version supported
]
```

Basically  you have each version and it's download URLs.  The meta data can then be used by the scripts for display purposes (the 'default' would be the current default distribution to download, the title is something to display, the category describes stable/milestone/snapshot and the display order can control the ordering in the UI).

Thoughts?

FYI - I can certainly submit a PR for this file (assuming against master not 2.5.x) and then followup with PR on the windows update scripts (and I'll change the windows installation [@kaikreuzer we still need to talk about that one btw]) but would like someone to do the linux side and pull in whoever does the openhabian side as well..

**Comment by wborn** [comment]
It's certainly useful and I would also adapt the Docker container build to use it. Due to the lack of such a file we now have to maintain our own [versions](https://github.com/openhab/openhab-docker/blob/master/versions) file in that repo.

It should also contain the milestone/RC versions and for 2.x versions also the URLs to legacy add-ons KAR.

**Comment by kaikreuzer** [comment]
IIRC, we also already once had a discussion with @ghys to have such a file for filling https://www.openhab.org/download/ with the latest infos - this might bring in some additional requirements.

**Comment by tmrobert8** [comment]
Thanks @wborn - @kaikreuzer I guess the first thing we need to decide is where it's hosted.  If you want it on that link - I definitely need to get some input from whomever is responsible for that (@ghys?).  

Second issue is how many versions back do we want it in?  Are you wanting all the way back and including the 1.x versions and beyond?

Note: I'm going to be gone for the next three days (in about 4 hours) and will respond when I get back..

**Comment by ghys** [comment]
Having a list of version could indeed be a valuable addition. 👍 

> we also already once had a discussion with @ghys to have such a file for filling https://www.openhab.org/download/ with the latest infos - this might bring in some additional requirements.

I vaguely remember this as well, but not exactly what these requirements were; apart from determining the latest version by release channel, I guess the same use cases as described above apply - mostly having a reliable source of valid versions and download links. We could also offer options to download previous versions of the distribution on the website too, if necessary. This will probably necessary at some point anyways, since we'd want to offer download links for the last 2.x release alongside 3.x builds - but the download page will likely get an overhaul then.

We could host such a file on https://www.openhab.org/versions.json or a similar well-known location, the website's build script could handle that, the real question is which repo if would come from (openhab-distro doesn't seem like a bad choice), and who's responsible for maintaining it and keeping it current, if it can't be automated.

As for the format and content, JSON seems fine and easily parsed, and the example by @tmrobert8 would work too. Maybe we could include the list of warnings and breaking changes as they're displayed by the scripts when upgrading, and URLs to the release notes/related community post.

**Comment by tmrobert8** [comment]
There is an update.lst file already in distro that describes the various changes from one version to the next.  I know both update scripts already process this one.  

If you don't mind waiting until monday - I can create a file of all 2.x and the 3.x versions and create a PR to have it sit somewhere in distro - @kaikreuzer have a preference to where?

**Comment by wborn** [comment]
To push this forwards I've generated a [versions.json](https://github.com/wborn/openhab-versions/blob/master/versions.json) file using a [bash script](https://github.com/wborn/openhab-versions) inspired by what we use in the Docker repo. WDYT?

**Comment by tmrobert8** [comment]
Sorry - real life and the darn sony addon has been taking up my time.  The versions.json is exactly what I was thinking.  We just need to find a home for it now...
